crates/burn-collective/src/local/all_reduce/base.rs

```diff
@@ -12,200 +14,90 @@ use crate::{
-        tensors: &mut HashMap<PeerId, B::FloatTensorPrimitive>,
-        op: ReduceOperation,
-        config: &CollectiveConfig,
-        global_client: &mut Node<B, WebSocket>,
-    ) -> Result<(), CollectiveError> {
-        let local_strategy = config.local_all_reduce_strategy;
-        let global_strategy = config.global_all_reduce_strategy;
-
-        // Get corresponding devices for each peer
-        let devices = tensors
-            .iter()
-            .map(|(id, tensor)| (*id, B::float_device(tensor)))
-            .collect::<HashMap<PeerId, B::Device>>();
-
-        // For Centralized and Tree, we only need to do a reduce here, we'll do a broadcast later
-        let main_device = *tensors.keys().next().unwrap();
-        let mut tensors_to_reduce = core::mem::take(tensors);
-        let mut main_tensor = match local_strategy {
-            AllReduceStrategy::Centralized => {
-                reduce_sum_centralized::<B>(tensors_to_reduce, &main_device)
-            }
-            AllReduceStrategy::Tree(arity) => {
-                reduce_sum_tree::<B>(tensors_to_reduce, &main_device, arity)
-            }
-            AllReduceStrategy::Ring => {
-                all_reduce_sum_ring::<B>(&mut tensors_to_reduce);
-                tensors_to_reduce.remove(&main_device).unwrap()
-            }
-        };
-
-        // Do aggregation on global level with the main tensor
-        main_tensor = global_client
+/// Do an all-reduce in a multi-node context
+///
+/// With Tree and Centralized strategies, the all-reduce is split between a
+/// reduce (all tensors are reduced to one device), and a broadcast (the result is sent to all
+/// other devices). The all-reduce on the global level is done between both steps.
+/// Due to the nature of the Ring strategy, this separation can't be done.
+// For the Ring strategy, this isn't possible, because it is more like a
+// reduce-scatter plus an all-gather, so using a Ring strategy locally in a multi-node
+// setup may be unadvantageous.
+#[tracing::instrument(skip(tensors, config, global_client))]
+pub(crate) async fn all_reduce_with_global<B: Backend>(
+    tensors: CollectiveTensorMap<B>,
+    op: ReduceOperation,
+    config: &CollectiveConfig,
+    global_client: &mut Node<B, WebSocket>,
+) -> Result<CollectiveTensorMap<B>, CollectiveError> {
+    let local_strategy = config.local_all_reduce_strategy;
+    let global_strategy = config.global_all_reduce_strategy;
+
+    // Get corresponding devices for each peer
+    let devices = tensors
+        .iter()
+        .map(|(id, tensor)| (*id, B::float_device(tensor)))
+        .collect::<HashMap<PeerId, B::Device>>();
+
+    // For Centralized and Tree, we only need to do a reduce here, we'll do a broadcast later
+    let main_device = *tensors.keys().next().unwrap();
+
+    let mut main_tensor = match local_strategy {
+        AllReduceStrategy::Centralized => reduce_sum_centralized::<B>(tensors, &main_device),
```

Comment:
It's strange that the global all reduce call `reduce_sum_centralized` and the local one call `all_reduce_sum_centralized`. When prefix by `all` it's local, and if not it's global?.

Maybe we should add proper pre/post fix?

Reply:
`reduce_sum(...) -> Tensor` and `all_reduce_sum() -> { PeerId: Tensor }` are different operations; but I do agree that there's duplication, and we probably want a local op-combinator library to avoid that duplication, even if those ops aren't shared to users.